### PR TITLE
Signal propagation plot

### PR DESCRIPTION
**Pull request by RisaMagpie**

### CODE REVIEW

File: batchflow/models/torch/visualization.py [hunk 1]
```diff
@@ -287,6 +287,58 @@ def get_gradcam(self, *args, targets=None, feed_dict=None,
...
+            'Average Channel Variance': []
+        }
+        modules_instances = []
+        
+        if len(image_shape) != 3:
+            raise ValueError("Parameter image_shape must be a sequence of 3 elements (C, H, W)")
+        image_shape = (batch_size, *image_shape)
+
+        try:
+            hooks_controller = ModelForwardExtractor(model)
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/visualization.py:309]: You can definitely re-use the same class with hooks without implementing your own

File: batchflow/models/torch/visualization.py [hunk 2]
```diff
@@ -287,6 +287,58 @@ def get_gradcam(self, *args, targets=None, feed_dict=None,
...
+
+        Under the hood, forward hooks are registered to capture outputs of all layers,
+        and they are removed after extraction of all the activations.
+        
+        ## Link and params"""
+        statistics = {
+            'Average Channel Squared Mean': [], 
+            'Average Channel Variance': []
+        }
+        modules_instances = []
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/visualization.py:302]: What is the reason for this list here?
Anyways, you can just make it as an additional key in your dictionary

File: batchflow/models/torch/visualization.py [hunk 3]
```diff
@@ -287,6 +287,58 @@ def get_gradcam(self, *args, targets=None, feed_dict=None,
...
+
+        try:
+            hooks_controller = ModelForwardExtractor(model)
+            _ = model(torch.randn(image_shape))
+        finally:
+            hooks_controller.close()
+
+        modules_instances = hooks_controller.modules_instances        
+        for tensor in hooks_controller.signals:
+            avg_ch_squared_mean = torch.mean(torch.mean(tensor, axis=1) ** 2).item()
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/visualization.py:316]: You can directly append this to stats: that would eliminate the need for strange variable names

File: batchflow/models/torch/visualization.py [hunk 4]
```diff
@@ -287,6 +287,58 @@ def get_gradcam(self, *args, targets=None, feed_dict=None,
...
+        """ Visualize signal propagation plot. """
+
+        if (statistics is None) and ((model is None) or (image_shape is None)):
+            raise ValueError("You should define statistics or model and image_shape")
+        else:
+            statistics, _ = get_signal_propagation(self, model=model, image_shape=image_shape, batch_size=batch_size)
+
+        fig, axes = plt.subplots(1, len(statistics), figsize=(15, 5))
+        for (ax, (title, data)) in zip(axes, statistics.items()):
+            x = range(len(data))
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/visualization.py:334]: Is that necessary?

File: batchflow/models/torch/visualization.py [hunk 5]
```diff
@@ -287,6 +287,58 @@ def get_gradcam(self, *args, targets=None, feed_dict=None,
...
+            ax.plot(x, data)
+            ax.set_title(title + " over network units", fontsize=14)
+            ax.set_xlabel("Number of network unit", fontsize=12)
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/visualization.py:337]: Network depth?

**Review by SergeyTsimfer — CHANGES_REQUESTED** [review]

File: batchflow/models/torch/visualization.py [hunk 6]
```diff
@@ -287,6 +287,96 @@ def get_gradcam(self, *args, targets=None, feed_dict=None,
...
+        <https://arxiv.org/pdf/2101.08692.pdf>`_"
+
+        Under the hood, forward hooks are registered to capture outputs of all layers,
+        and they are removed after extraction of all the activations.
+
+        Parameters
+        ----------
+        model : nn.Module
+            Model to base visualizations on.
+        input_shape : sequence
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/visualization.py:303]: You can use the last model input shape

File: batchflow/models/torch/visualization.py [hunk 7]
```diff
@@ -287,6 +287,96 @@ def get_gradcam(self, *args, targets=None, feed_dict=None,
...
+            Shape of the image to generate.
+        batch_size : int
+            Number of images in one batch.
+        input : Tensor
+            Input tensor for signal propagation.
+        """
+        if model is None:
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/visualization.py:310]: `model = model or self.model` is enough

File: batchflow/models/torch/visualization.py [hunk 8]
```diff
@@ -287,6 +287,96 @@ def get_gradcam(self, *args, targets=None, feed_dict=None,
...
+            'Modules instances': []
+        }
+        extractors = []
+        signals = []
+
+        try:
+            for module in model.modules():
+                submodules_amount = sum(1 for _ in module.modules())
+                if submodules_amount == 1:
+                    module_instance = module.__class__.__name__
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/visualization.py:336]: move this line above and merge two `if` checks

File: batchflow/models/torch/visualization.py [hunk 9]
```diff
@@ -287,6 +287,96 @@ def get_gradcam(self, *args, targets=None, feed_dict=None,
...
+                    if module_instance != 'Identity':
+                        statistics['Modules instances'].append(module_instance)
+                        extractors.append(LayerExtractor(module))
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/visualization.py:339]: This is more important than the previous line; also, you've created `extractors` and `signals` in other order. So, move this line up

File: batchflow/models/torch/visualization.py [hunk 10]
```diff
@@ -287,6 +287,96 @@ def get_gradcam(self, *args, targets=None, feed_dict=None,
...
+        ----------
+        kwargs : dict
+            Named arguments directly passed to `get_signal_propagation` method.
+            `statistics` or `model` and `input_shape` must be provided.
+        """
+        statistics = kwargs.get('statistics')
+        model = kwargs.get('model')
+        input_shape = kwargs.get('input_shape')
+        batch_size = kwargs.get('batch_size', 3)
+        input_tensor = kwargs.get('input_tensor')
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/visualization.py:367]: Don't do that, please

File: batchflow/models/torch/visualization.py [hunk 11]
```diff
@@ -287,6 +287,96 @@ def get_gradcam(self, *args, targets=None, feed_dict=None,
...
+
+        for tensor in signals:
+            avg_ch_squared_mean = torch.mean(torch.mean(tensor, axis=1) ** 2).item()
+            avg_ch_var =  torch.mean(torch.var(tensor, axis=1)).item()
+
+            statistics['Average Channel Squared Mean'].append(avg_ch_squared_mean)
+            statistics['Average Channel Variance'].append(avg_ch_var)
+        return statistics
+
+    def show_signal_propagation_plot(self, **kwargs):
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/visualization.py:354]: Make it `get_signal_propagation_plot`

File: batchflow/models/torch/visualization.py [hunk 12]
```diff
@@ -287,6 +287,96 @@ def get_gradcam(self, *args, targets=None, feed_dict=None,
...
+        input_tensor = kwargs.get('input_tensor')
+
+        if statistics is None:
+            statistics = self.get_signal_propagation(model=model, input_shape=input_shape,
+                                                     batch_size=batch_size, input_tensor=input_tensor)
+
+        fig, axes = plt.subplots(1, len(statistics)-1, figsize=(15, 5))
+        for (ax, (title, data)) in zip(axes, statistics.items()):
+            ax.plot(data)
+            ax.set_title(title + " over network units", fontsize=14)
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/visualization.py:376]: Add `grid` to the plot

**Review by SergeyTsimfer — APPROVED** [review]

**Comment by roman-kh** [comment]
Status check failed. Fix it before merging.

**Pull request merged**